File: src/partial.cpp
// File:    partial.cpp
// Purpose: Partial class
// Author:  Stephen Gold sgold@sonic.net
// (c) Copyright 2012 Stephen Gold
// Distributed under the terms of the GNU General Public License

/*
This file is part of the Gold Tile Game.

The Gold Tile Game is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by the 
Free Software Foundation, either version 3 of the License, or (at your 
option) any later version.

The Gold Tile Game is distributed in the hope that it will be useful, but 
WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY 
or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License 
for more details.

You should have received a copy of the GNU General Public License
along with the Gold Tile Game.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "game.hpp"
#include "partial.hpp"


// lifecycle

Partial::Partial(Game const *pGame, HintType strength) {
    Reset(pGame, strength);
}
// Partial(Partial const &);  compiler-generated copy constructor is OK
// ~Partial(void);  compiler-generated destructor is OK

void Partial::Reset(Game const *pGame, HintType strength) {
	mpGame = pGame;
    mHintStrength = strength;
    Reset();
}

// method invoked by takeback and at the start of a turn
void Partial::Reset(void) {
    mActiveId = Tile::ID_NONE;
    mHintedCellsValid = false;
    mPlayedTileCnt = 0;
	mSwapIds.MakeEmpty();
	if (HaveGame()) {
        mBoard = Board(*mpGame);
        mTiles = mpGame->ActiveTiles();
    } else {
	    mBoard.MakeEmpty();
		mTiles.MakeEmpty();
	}
}


// operators

// BaseBoard &operator=(BaseBoard const &);  compiler-generated assignment operator is OK

Partial::operator Board(void) const {
   return mBoard;
 }

Partial::operator HintType(void) const {
	return mHintStrength;
}


// misc methods

void Partial::Activate(TileIdType id) {
    ASSERT(id != Tile::ID_NONE);
    ASSERT(mActiveId == Tile::ID_NONE);
    ASSERT(Contains(id));

    mActiveId = id;     
    mHintedCellsValid = false;
}

// Add cells from rBasis to the hint-set if they are valid next
// steps for the given Tile.
void Partial::AddValidNextUses(
    Move const &rMove, 
    Tile const &rTile, 
    Cells const &rBasis)
{
    Cells::ConstIterator i_cell;
    for (i_cell = rBasis.begin(); i_cell != rBasis.end(); i_cell++) {
        Cell cell = *i_cell;
        if (!mHintedCells.Contains(cell) 
         && IsValidNextStep(rMove, cell, rTile))
        {
      		mHintedCells.Add(cell);
        }
    }
}

void Partial::BoardToHand(void) {
    ASSERT(mActiveId != Tile::ID_NONE);
    ASSERT(IsOnBoard(mActiveId));
    ASSERT(!IsInHand(mActiveId));
    
    Cell cell = LocateTile(mActiveId);
    mBoard.MakeEmpty(cell);
    --mPlayedTileCnt;
    mHintedCellsValid = false;

	ASSERT(mBoard.HasEmptyCell(cell));
    ASSERT(!IsOnBoard(mActiveId));
    ASSERT(IsInHand(mActiveId));
}

unsigned Partial::CountHand(void) const {
    ASSERT(CountTiles() >= CountPlayed() + CountSwap());
    
    unsigned result = CountTiles() - CountPlayed() - CountSwap();
    
    return result; 
}

unsigned Partial::CountHinted(void) {
    if (!mHintedCellsValid) {
        SetHintedCells();
    }
    ASSERT(mHintedCellsValid);

    unsigned result = mHintedCells.Count();

	return result;
}

unsigned Partial::CountPlayed(void) const {
    return mPlayedTileCnt; 
}

unsigned Partial::CountSwap(void) const {
    unsigned result = mSwapIds.Count();
    
    ASSERT(result <= CountTiles());
    ASSERT(!HaveGame() || result <= mpGame->CountStock());

    return result; 
}

unsigned Partial::CountTiles(void) const {
    unsigned result = mTiles.Count();
    
    return result; 
}

void Partial::Deactivate(void) {
    if (mActiveId != Tile::ID_NONE) {
        mActiveId = Tile::ID_NONE;     
        mHintedCellsValid = false;
    }
}

Cell Partial::FirstHinted(void) {
    if (!mHintedCellsValid) {
        SetHintedCells();
    }
    ASSERT(mHintedCellsValid);
	ASSERT(!mHintedCells.IsEmpty());
    Cells::ConstIterator i_cell = mHintedCells.begin();
	Cell result = *i_cell;

	return result;
}

GameStyleType Partial::GameStyle(void) const {
    GameStyleType result = GAME_STYLE_NONE;
    if (HaveGame()) {
        result = mpGame->Style();
    }
    
    return result;
}

TileIdType Partial::GetActive(void) const {
    return mActiveId;
}

TileIdType Partial::GetCellTile(Cell const &rCell) const {
    TileIdType result = Tile::ID_NONE;
    
    Tile const *p_tile = mBoard.GetCell(rCell);
    if (p_tile != NULL) {
        result = p_tile->Id();
    }

    return result;
}
     
Move Partial::GetMove(bool includeActiveFlag) const {
    Move result;
    
    for (unsigned i = 0; i < CountTiles(); i++) {
        Tile tile = GetTileByIndex(i);
        TileIdType id = tile.Id();
		if (includeActiveFlag || !IsActive(id)) {
            Cell cell;
            if (mBoard.LocateTileId(id, cell)) {
                result.Add(tile, cell);
            } else if (IsInSwap(id)) {
                result.Add(tile);
			}
        }
    }

	return result;
}

Tile Partial::GetTileById(TileIdType id) const {
    ASSERT(id != Tile::ID_NONE);

    Tile result;
     
    if (Contains(id)) {
        result = mTiles.FindTile(id);
    } else {
        Cell cell;
        bool found = mBoard.LocateTileId(id, cell);
        ASSERT(found);
        Tile const *p_tile = mBoard.GetCell(cell);
        ASSERT(p_tile != NULL);
        result = *p_tile;
    }
    
    ASSERT(result.Id() == id); 
    return result;                       
}

Tile Partial::GetTileByIndex(unsigned i) const {
    ASSERT(i < mTiles.Count());

    Tile result = mTiles[i];
    
    return result;                       
}

void Partial::HandToCell(Cell const &cell) {
    ASSERT(mActiveId != Tile::ID_NONE);
	ASSERT(IsInHand(mActiveId));
	ASSERT(mBoard.HasEmptyCell(cell));
	ASSERT(!IsOnBoard(mActiveId));

    Tile tile = mTiles.FindTile(mActiveId);
    mBoard.PlayOnCell(cell, tile);
    ++mPlayedTileCnt;
    mHintedCellsValid = false;
    
	ASSERT(!IsInHand(mActiveId));
    ASSERT(mBoard.GetCell(cell) != NULL);
    ASSERT(*(mBoard.GetCell(cell)) == tile);
}

void Partial::HandToSwap(void) {
    ASSERT(mActiveId != Tile::ID_NONE);
	ASSERT(!mSwapIds.Contains(mActiveId));
	
	mSwapIds.Add(mActiveId);
	mHintedCellsValid = false;
	
	ASSERT(mSwapIds.Contains(mActiveId));
}

Cell Partial::LocateTile(TileIdType id) const {
    Cell result;
    bool success = mBoard.LocateTileId(id, result);
    ASSERT(success);
    
    return result;
}

void Partial::SetHintedCells(void) {
    ASSERT(!mHintedCellsValid);
    ASSERT(HaveGame());

	// start with no cells hinted
    mHintedCells.MakeEmpty();
	if (mHintStrength == HINT_NONE) {
        mHintedCellsValid = true;
		return;
	}

    // hint all valid empty cells (from start of turn)
	int row_fringe = 1;
	int column_fringe = 1;
	if (Cell::Grid() == GRID_HEX) {
		row_fringe = 2;
	}
    IndexType top_row = row_fringe + mBoard.NorthMax();
    IndexType bottom_row = -row_fringe - mBoard.SouthMax();
    IndexType right_column = column_fringe + mBoard.EastMax();
    IndexType left_column = -column_fringe - mBoard.WestMax();
    ASSERT(bottom_row <= top_row);
    ASSERT(left_column <= right_column);
    for (IndexType row = top_row; row >= bottom_row; row--) {
        for (IndexType column = left_column; column <= right_column; column++) {
			if (Cell::IsValid(row, column)) {
                Cell cell(row, column);
                if (mpGame->HasEmptyCell(cell)) {
                    mHintedCells.Add(cell);
                }
			}
        }
    }
	if (mHintStrength == HINT_EMPTY) {
        mHintedCellsValid = true;
		return;
	}
    
    // unhint any cells that are not the start or the neighbor of a used cell
	Cells base = mHintedCells;
	mHintedCells.MakeEmpty();
	if (mBoard.IsEmpty()) {
        Cell start;
        mHintedCells.Add(start);
    } else {
        Cells::Iterator i_cell;
        for (i_cell = base.begin(); i_cell != base.end(); i_cell++) {
            Cell cell = *i_cell;
            if (mBoard.HasNeighbor(cell)) {
                mHintedCells.Add(cell);
            }
        }
    }
	if (mHintStrength == HINT_CONNECTED) {
        mHintedCellsValid = true;
		return;
	}

	// unhint cells not usable ...
	//   HINT_USABLE_ANY:       by any available tile
	//   HINT_USABLE_SELECTED:  by the selected tile

    base = mHintedCells;
    mHintedCells.MakeEmpty();
    Move move = GetMove(false);

	for (unsigned i = 0; i < CountTiles(); i++) {
        Tile tile = mTiles[i];
	    TileIdType id = tile.Id();
        bool include_tile = (!mBoard.ContainsId(id) || id == mActiveId);
        if (mHintStrength == HINT_USABLE_SELECTED && mActiveId != Tile::ID_NONE) {
            include_tile = (mActiveId == id);
        }
		if (include_tile) {
            AddValidNextUses(move, tile, base);
        }
    }
    
    mHintedCellsValid = true;
}

void Partial::SetHintStrength(HintType strength) {
	switch (strength) {
	    case HINT_NONE:
		case HINT_EMPTY:
		case HINT_CONNECTED:
		case HINT_USABLE_ANY:
		case HINT_USABLE_SELECTED:
		    break;
		default:
			FAIL();
	}

	if (mHintStrength != strength) {
        mHintedCellsValid = false;
	}
    mHintStrength = strength;
}

void Partial::SwapAll(void) {
    ASSERT(mActiveId == Tile::ID_NONE);
    ASSERT(CanSwapAll());

	for (unsigned i = 0; i < CountTiles(); i++) {
        Tile tile = mTiles[i];
	    TileIdType id = tile.Id();
		if (!mSwapIds.Contains(id)) {
	        mSwapIds.Add(id);
		}
	}
}

void Partial::SwapToHand(void) {
    ASSERT(mActiveId != Tile::ID_NONE);
	ASSERT(mSwapIds.Contains(mActiveId));
	
	mSwapIds.Remove(mActiveId);
    mHintedCellsValid = false;
	
	ASSERT(!mSwapIds.Contains(mActiveId));
}

// inquiry methods

bool Partial::CanSwapAll(void) const {
	bool result = false;
    if (HaveGame()) {
        result = (mpGame->CountStock() >= CountTiles());
    }

	return result;
}

bool Partial::Contains(TileIdType id) const {
     bool result = mTiles.ContainsId(id);     

     return result;
}

bool Partial::HaveGame(void) const {
     bool result = (mpGame != NULL);
     
     return result;
}

bool Partial::IsActive(TileIdType id) const {
    ASSERT(id != Tile::ID_NONE);
    bool result = (mActiveId == id);
    
    return result;
}

bool Partial::IsEmpty(Cell const &rCell) const {
    Tile const *p_tile = mBoard.GetCell(rCell);
    bool result = (p_tile == NULL);

    return result;
}

bool Partial::IsGameOver(void) const {
     bool result = false;
     if (HaveGame()) {
         result = mpGame->IsOver();
     }
     
     return result;
}

bool Partial::IsGamePaused(void) const {
	bool result = false;
	if (HaveGame() && !IsGameOver()) {
		result = mpGame->IsPaused();
	}

	return result;
}

bool Partial::IsHinted(Cell const &rCell) {
    if (!mHintedCellsValid) {
        SetHintedCells();
    }
    ASSERT(mHintedCellsValid);
    
	bool result = mHintedCells.Contains(rCell);
	
	return result;
}

bool Partial::IsInHand(TileIdType id) const {
    bool result = true;
    
    if (!mTiles.ContainsId(id)) {
        result = false;
    } else if (IsOnBoard(id)) {
        result = false;
    } else if (IsInSwap(id)) {
        result = false;
    }
    
    return result;
}

bool Partial::IsInSwap(TileIdType id) const {
    bool result = mSwapIds.Contains(id);
    
    return result;
}

bool Partial::IsOnBoard(TileIdType id) const {
    bool result = mBoard.ContainsId(id);
    
    return result;
}

bool Partial::IsPass(void) const {
    bool result = (CountPlayed() + CountSwap() == 0);

    return result;
}

bool Partial::IsValidNextStep(
	Move const &rBase,
	Cell const &rCell,
	Tile const &rTile) const
{
	// Check whether a hypothetical next step would be legal.
	ASSERT(HaveGame());

	Move move = rBase;
	move.Add(rTile, rCell);
    char const *reason;
	bool result = mpGame->IsLegalMove(move, reason);

	if (!result && ::str_eq(reason, "FIRST")) {  
		// legal as a partial move.
	    result = true;
	}

	return result;
}

bool Partial::IsVisible(Cell const &rCell) {
	int fringe = 1;
	if (Cell::Grid() == GRID_HEX) {
		fringe = 2;
	}

    IndexType top_row = fringe + mBoard.NorthMax();
    IndexType bottom_row = -fringe - mBoard.SouthMax();
    IndexType right_column = fringe + mBoard.EastMax();
    IndexType left_column = -fringe - mBoard.WestMax();
    ASSERT(bottom_row <= top_row);
    ASSERT(left_column <= right_column);

	IndexType row = rCell.Row();
	IndexType column = rCell.Column();

	bool result = (row >= bottom_row) 
		       && (row <= top_row)
			   && (column >= left_column)
			   && (column <= right_column);

	return result;
}

File: trunk/Java/GoldTile/src/goldtile/Partial.java
// File:     Partial.java
// Location: Java/GoldTile/src/goldtile
// Purpose:  Partial class for the Gold Tile Game
/**
 * @author Stephen Gold
 */
// (c) Copyright 2012 Stephen Gold
// Distributed under the terms of the GNU General Public License

/*
This file is part of the Gold Tile Game.

The Gold Tile Game is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by the
Free Software Foundation, either version 3 of the License, or (at your
option) any later version.

The Gold Tile Game is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
for more details.

You should have received a copy of the GNU General Public License
along with the Gold Tile Game.  If not, see <http://www.gnu.org/licenses/>.
*/

package goldtile;

public class Partial {
    // enums
    public static enum Active { EXCLUDED, INCLUDED };

    // constants
    final private static Fraction SUGGEST_SKIP_PROBABILITY = new Fraction(0.0);

    // per-instance fields (all mutable), sorted by type
    private Board board = new Board();
    private Cells hintedCells = null;       // cached data, null means invalid
    private Fraction skipProbability = null;
    private HintStrength hintStrength = null;
    private int playedTileCount = 0;        // cached data, keep updated
    private Integer pointCount = null;      // cached data, null means invalid
    private Tile activeTile = null;
    private Tiles playableTiles = new Tiles();
    private Tiles swapTiles = new Tiles();

    // constructors

    public Partial() {
        takeBackTiles();
        hintStrength = HintStrength.NONE;
        skipProbability = HandOpt.SKIP_PROBABILITY_DEFAULT;
    }

    private Partial(Partial other) {
        copy(other);
    }

    // methods, sorted by name

    public void activate(Tile tile) {
        assert isPlayable(tile);
        assert Game.hasInstance();

        if (activeTile == null || !tile.equals(activeTile)) {
            activeTile = tile;

            hintedCells = null;
            // no effect on playedTileCount
            // no effect on pointCount
        }
    }

    /**
     * Add cells from "base" to "hintedCells" if they are valid
     * uses for the specified Tile.
     */
    private void addValidUses(ReadMove move, Tile tile, Cells base) {
        assert move != null;
        assert tile != null;
        assert base != null;

        for (Cell cell : base) {
            if (!hintedCells.contains(cell) &&
                    isValidNextStep(move, cell, tile))
            {
                hintedCells.add(cell);
            }
        }
    }

    public void assertActive(Place expectedPlace) {
        final Place actualPlace = findActiveTile();
        assert actualPlace.equals(expectedPlace) : actualPlace;
    }

    public ReadMove autoPlay() {
        assert Game.hasInstance();
        final ReadHandOpt handOpt = Game.getInstance().getPlayable().getOpt();
        assert handOpt.isAutomatic();

        final HintStrength saveStrength = hintStrength;
        final Fraction probability = handOpt.getSkipProbability();
        ReadMove result;
        try {
            final Partial best = findBestMove(probability);
            result = best.getMove(Partial.Active.INCLUDED);
            assert Game.getInstance().checkMove(result) == null;
        } catch (InterruptedException exception) {
            result = new Move(playableTiles);  // resign
        }

        return result;
    }

    private void boardToHand() {
        assertActive(Place.BOARD);

        final Cell cell = board.find(activeTile);
        board.clear(cell);

        hintedCells = null;
        --playedTileCount;
        pointCount = null;

        assertActive(Place.HAND);
    }

    public boolean canSwapAll() {
        return Game.hasInstance() &&
                Game.getInstance().countStock() >= countPlayableTiles();
        // note: doesn't consider Game.mustPlay
    }

    public void changeHand() {
        final ReadHand playable = Game.getInstance().getPlayable();
        final ReadHandOpt handOpt = playable.getOpt();
        skipProbability = handOpt.getSkipProbability();

        takeBackTiles();
    }

    final public void copy(Partial other) {
        board = new Board(other.board);
        hintedCells = (other.hintedCells == null) ? null :
                new Cells(other.hintedCells);
        skipProbability = other.skipProbability;
        hintStrength = other.hintStrength;
        playedTileCount = other.playedTileCount;
        pointCount = (other.pointCount == null) ? null :
                new Integer(other.pointCount);
        activeTile = other.activeTile;
        playableTiles = new Tiles(other.playableTiles);
        swapTiles = new Tiles(other.swapTiles);
    }

    public int countHandTiles() {
        assert countPlayableTiles() >= countPlayedTiles() + countSwappedTiles()
                : countPlayedTiles();

        return countPlayableTiles() - countPlayedTiles() - countSwappedTiles();
    }

    public int countHintedCells() {
        if (hintedCells == null) {
            updateHintedCells();
            assert hintedCells != null;
        }
        return hintedCells.size();
    }

    public int countPlayableTiles() {
        return playableTiles.size();
    }

    public int countPlayedTiles() {
        return playedTileCount;
    }

    public int countSwappedTiles() {
        return swapTiles.size();
    }

    public Tile deactivate() {
        assert Game.hasInstance();

        final Tile result = activeTile;

        if (activeTile != null) {
            activeTile = null;

            hintedCells = null;
            // no effect on playedTileCount
            // no effect on pointCount
        }

        return result;
    }

    public Place find(Tile tile) {
        if (isInSwap(tile)) {
            assert !board.contains(tile) : board.find(tile);
            return Place.SWAP_AREA;
        } else if (board.contains(tile)) {
            return Place.BOARD;
        } else if (isPlayable(tile)) {
            return Place.HAND;
        } else {
            return null;
        }
    }

    public Place findActiveTile() {
        return find(activeTile);
    }

    static public Partial findBestMove(Fraction skipProbability)
            throws InterruptedException
    {
        assert skipProbability != null;
        assert Game.hasInstance();

        if (Game.getInstance().getMustPlay() > 0) {
            // Consider ALL possibile moves.
            skipProbability = new Fraction(0.0);
        }

        final Partial partial = new Partial();
        partial.takeBackTiles();
        partial.setSkipProbability(skipProbability);
        partial.setHintStrength(HintStrength.USABLE_BY_ACTIVE);

        Partial result = new Partial(partial); // a temporary copy

        // Search for good plays.
        result = partial.findBestPlay(result);
        assert result.playedTileCount >= Game.getInstance().getMustPlay();

        if (result.getPointCount() == 0) {
            // Construct a swap.
            result.takeBackTiles();

            int swapCount = result.countPlayableTiles();
            final int maxSwap = Game.getInstance().countStock();
            if (swapCount > maxSwap) {
                swapCount = maxSwap;
            }

            // TODO be more selective in choosing tiles to swap
            for (Tile tile : result.playableTiles) {
                if (swapCount <= 0) {
                    break;
                }
                result.activate(tile);
                result.handToSwap();
                swapCount--;
            }
        }

        // note: result may incorporate an active tile

        return result;
    }

    // RECURSIVE
    private Partial findBestPlay(Partial bestSoFar)
            throws InterruptedException
    {
        assert bestSoFar != null;
        assert Game.hasInstance();

        Partial result = bestSoFar;
        if (getPointCount() > bestSoFar.getPointCount()) {
            result.copy(this);
        }

        final Tile saveActive = deactivate();

        for (Tile tile : playableTiles) {
            if (isInHand(tile) && !skipProbability.randomBoolean()) {
                if (Thread.interrupted()) {
                    throw new InterruptedException();
                }
                activate(tile);
                updateHintedCells();
                for (Cell cell : hintedCells) {
                    if (Thread.interrupted()) {
                        throw new InterruptedException();
                    }
                    handToBoard(cell);
                    result = findBestPlay(bestSoFar);
                    boardToHand();
                }
                deactivate();
            }
        }

        if (saveActive != null) {
            activate(saveActive);
        }

        return result;
    }

    public Cell firstHintedCell() {
        if (hintedCells == null) {
            updateHintedCells();
            assert hintedCells != null;
        }
        assert hintedCells.size() > 0;

        return hintedCells.first();
    }

    public Tile getActiveTile() {
        return activeTile;
    }

    public ReadBoard getBoard() {
        return board;
    }

    public Tiles getBoardTiles() {
        final Tiles result = new Tiles();

        for (Tile tile : playableTiles) {
            if (board.contains(tile)) {
                result.add(tile);
            }
        }

        return result;
    }

    public HintStrength getHintStrength() {
        return hintStrength;
    }

    public ReadMove getMove(Active includeActive) {
        final Move result = new Move();

        for (Tile tile : playableTiles) {
            if (includeActive == Active.INCLUDED || !isActive(tile)) {
                final Cell destination = board.find(tile);
                if (destination != null || isInSwap(tile)) {
                    result.add(tile, destination);
                }
            }
        }

        return result;
    }

    public int getPointCount() {
        if (pointCount == null) {
            updatePointCount();
            assert pointCount != null;
        }

        return pointCount;
    }


    public void handToBoard(Cell cell) {
        assert cell != null;
        assert cell.isValid();
        assert board.isEmpty(cell);
        assertActive(Place.HAND);

        board.place(cell, activeTile);

        hintedCells = null;
        playedTileCount++;
        pointCount = null;

        assertActive(Place.BOARD);
    }

    public void handToSwap() {
        assertActive(Place.HAND);

        swapTiles.add(activeTile);

        hintedCells = null;
        // no effect on playedTileCount
        // no effect on pointCount

        assertActive(Place.SWAP_AREA);
    }

    public boolean hasActiveTile() {
        return activeTile != null;
    }

    public boolean isActive(Tile tile) {
        if (tile == null) {
            return activeTile == null;
        } else {
            return tile.equals(activeTile);
        }
    }

    public boolean isHinted(Cell cell) {
        assert cell != null;

        if (hintedCells == null) {
            updateHintedCells();
            assert hintedCells != null;
        }

        return hintedCells.contains(cell);
    }

    public boolean isInHand(Tile tile) {
        if (tile == null) {
            return false;
        } else {
            return isPlayable(tile) && !isInSwap(tile) && !board.contains(tile);
        }
    }

    public boolean isInSwap(Tile tile) {
        if (tile == null) {
            return false;
        } else {
            return swapTiles.contains(tile);
        }
    }

    public boolean isPass() {
        return countPlayedTiles() + countSwappedTiles() == 0;
    }

    public boolean isPlayable(Tile tile) {
        if (tile == null) {
            return false;
        } else {
            return playableTiles.contains(tile);
        }
    }

    /**
     * Check whether a hypothetical next step would be a legal
     * partial move.
     * @param base
     * @param cell
     * @param tile
     * @return true or false
     */
    private boolean isValidNextStep(ReadMove base, Cell cell, Tile tile) {
        assert base != null;
        assert cell != null;
        assert tile != null;
        assert Game.hasInstance();

        final Move move = new Move(base);
        move.add(tile, cell);
        final UserMessage reason = Game.getInstance().checkMove(move);

        return reason == null || reason == UserMessage.FIRST_TURN;
    }

    public void moveActiveTile(Place from, Place to, Cell toCell) {
        assert from != null;
        assert to != null;
        assert isPlayable(activeTile);

        // Move tile to hand -- temporarily in many cases.
        switch (from) {
            case BOARD:
                boardToHand();
                break;

            case SWAP_AREA:
                swapToHand();
                break;
        }

        // Move tile from hand to destination.
        switch (to) {
            case BOARD:
                assert toCell != null;
                assert toCell.isValid();

                handToBoard(toCell);
                break;

            case SWAP_AREA:
                handToSwap();
                break;
        }
    }

    public void nextHand() {
        Game.getInstance().nextHand();
        changeHand();
    }

    public void setHintStrength(HintStrength strength) {
        if (strength != null && !strength.equals(hintStrength)) {
            hintStrength = strength;
            hintedCells = null;
        }
    }

    public void setSkipProbability(Fraction skipProbability) {
        assert skipProbability != null;
        assert !skipProbability.isUnity();

        this.skipProbability = skipProbability;
    }

    public Partial suggest() {
        assert Game.hasInstance();
        assert Game.getInstance().getPlayable().getOpt().isLocalUser();

        final HintStrength saveStrength = hintStrength;

        Partial best;
        try {
            best = findBestMove(SUGGEST_SKIP_PROBABILITY);
            best.deactivate();
            best.setHintStrength(saveStrength);
        } catch (InterruptedException exception) {
            best = null;
        }

        return best;
    }

    public void swapAllTiles() {
        if (countSwappedTiles() < countPlayableTiles()) {
            takeBackTiles();
            swapTiles = new Tiles(playableTiles);
        }
    }

    private void swapToHand() {
        assertActive(Place.SWAP_AREA);

        swapTiles.remove(activeTile);
        hintedCells = null;
        // no effect on pointCount

        assertActive(Place.HAND);
    }

    final public void takeBackTiles() {
        activeTile = null;
        swapTiles.clear();

        hintedCells = null;
        playedTileCount = 0;
        pointCount = 0;

        if (Game.hasInstance()) {
            // TODO might be more efficient to modify the existing board copy
            board = Game.getInstance().copyBoard();
            playableTiles = Game.getInstance().getPlayable().copyContents();
        } else {
            board.clear();
            playableTiles.clear();
        }
    }

    private void updateHintedCells() {
        hintedCells = new Cells();
        if (hintStrength == HintStrength.NONE) {
            return;
        }

        // Add all valid empty cells.
        for (int row = board.getBottomUseRow();
                row <= board.getTopUseRow(); row++)
        {
            for (int column = board.getLeftUseColumn();
                    column <= board.getRightUseColumn(); column++)
            {
                final Cell cell = new Cell(row, column);
                final Cell wrapCell = cell.wrap();
                if (wrapCell.isValid() &&
                    board.isEmpty(wrapCell) &&
                    !hintedCells.contains(wrapCell))
                {
                    hintedCells.add(cell);
                }
            }
        }
        if (hintStrength == HintStrength.UNUSED) {
            return;
        }

        /*
         * Remove any cells that are neither the start cell nor
         * the neighbor of a used cell.
         */
        Cells base = new Cells(hintedCells);
        for (Cell cell : base) {
            if (!cell.isStart() && !board.hasUsedNeighbor(cell)) {
                hintedCells.remove(cell);
            }
        }
        if (hintStrength == HintStrength.CONNECTED) {
            return;
        }

        /*
         * Remove any cells that are not ...
         *   USABLE_BY_PLAYABLE:  usable by any playable tile
         *   USABLE_BY_ACTIVE:    usable by the active tile
         */
        base = new Cells(hintedCells);
        hintedCells.clear();
        final ReadMove move = getMove(Active.EXCLUDED);
        for (Tile tile : playableTiles) {
             boolean includeTile;
             if (hintStrength == HintStrength.USABLE_BY_ACTIVE && hasActiveTile()) {
                 includeTile = isActive(tile);
            } else {
                includeTile = (!board.contains(tile) || isActive(tile));
            }
            if (includeTile) {
                addValidUses(move, tile, base);
            }
        }
    }

    private void updatePointCount() {
        assert Game.hasInstance();

        pointCount = new Integer(0);

        final int mustPlay = Game.getInstance().getMustPlay();
        if (mustPlay == 0 || countPlayedTiles() >= mustPlay) {
            final ReadMove move = getMove(Active.INCLUDED);
            pointCount = board.score(move);
        }
    }
}
